File: ROID-Forum-Server/ROID-Forum-Server/Convert Server Code/GMGServer/bin/Server/Sections/Game Section Controller/ThreadController.dart
library GameThreadController;
import 'ThreadInfo.dart';
import 'GameSectionController.dart';
import '../../User.dart';
import 'CommentInfo.dart';

class ThreadController {
  List<ThreadInfo> threads;
  GameSectionController controller;
  int threadIDs = 0;
    ThreadController(GameSectionController c) {
      controller = c;
      threads = new List();
    }
    ThreadInfo getThreadForID(int threadID) {
      for (int i = 0; i < threads.length; i++) {
        if (threads[i].id == threadID) {
          return threads[i];
        }
      }
      return null;
    }
    //Thread actions
    void addThread(User p, String title, String description) {
      ThreadInfo t = new ThreadInfo.fromData(p.account.name, threadIDs++, title, description);
      threads.add(t);
      controller.messageSender.sendAddThreadToAll(t);

      //Send a message to the All Controller
      controller.server.allSection.threadController.addThread(controller.name, t.id, t.title, t.comments.length, t.owner);

      moveThreadToTop(t.id);
    }
    void deleteThread(User p,int threadID) {
      ThreadInfo t = getThreadForID(threadID);
      if (t != null) {
        if (t.owner == p.account.name) {
          threads.remove(t);
          controller.messageSender.sendRemoveThreadToAll(t);
          //Send a message to the All Controller
          controller.server.allSection.threadController.deleteThread(controller.name, t.id);
        }
      }
    }
    void editThread(User p,int id, String title, String description) {
      ThreadInfo t = getThreadForID(id);
      if (t != null) {
            if (t.owner == p.account.name) {
      t.title = title;
      t.description = description;
      controller.messageSender.sendUpdateThreadToAll(t);
      //Send a message to the All Controller
      controller.server.allSection.threadController.editThread(controller.name, t.id, t.title);
            }
      }
    }
    void moveThreadToTop(int id) {
      ThreadInfo t = getThreadForID(id);
      if (t != null) {
        threads.remove(t);
        threads.insert(0, t);
        controller.messageSender.sendMoveThreadToTopToAll(t);
        //Send a message to the All Controller
        controller.server.allSection.threadController.moveThreadToTop(controller.name, t.id);
      }
    }
    void addComment(User u, int threadID, String text) {
      ThreadInfo t = getThreadForID(threadID);
      if (t != null) {
        CommentInfo c = new CommentInfo.fromData(t.id, t.commentIDs++, text, u.account.name);
        t.comments.add(c);
        controller.messageSender.sendAddCommentToAll(c);

        //Send a message to the All Controller
        controller.server.allSection.threadController.updateCommentCount(controller.name, t.id, t.comments.length);

        moveThreadToTop(threadID);
      }
    }
    void deleteComment(User p,int threadID, int commentID) {
      ThreadInfo t = getThreadForID(threadID);
      if (t != null) {
        CommentInfo c = t.getCommentForID(commentID);
        if (c != null) {
          if (p.account.name == c.owner) {
          t.comments.remove(c);
          controller.messageSender.sendDeleteCommentToAll(c);

          //Send a message to the All Controller
          controller.server.allSection.threadController.updateCommentCount(controller.name, t.id, t.comments.length);
          }
        }
      }
    }
    void editComment(User p,int threadID, int commentID, String description) {
      ThreadInfo t = getThreadForID(threadID);
      if (t != null) {
        CommentInfo c = t.getCommentForID(commentID);
        if (c != null) {
          if (p.account.name == c.owner) {
            c.comment = description;
            controller.messageSender.sendUpdateCommentToAll(c);
          }
        }
      }
    }
  }

File: ROID-Forum-Server/ROID-Forum-Server/Section Controller/ThreadController.cs
using System;
using Cassandra;

namespace ROIDForumServer
{
    public class ThreadController(SectionController sectionController, ISession databaseSession)
    {
        private ISession DatabaseSession { get; } = databaseSession;
        public void AddThread(ConnectedUser user, string title, string description)
        {
            if (user.AccountId == null)
            {
                return;
            }

            Guid threadId = DatabaseThread.CreateThread(DatabaseSession, (Guid)user.AccountId, sectionController.SectionId, title);
            DatabaseComment.CreateComment(DatabaseSession, (Guid)user.AccountId, threadId, sectionController.SectionId, description);
            sectionController.MessageSender.SendAddThreadToAll(threadId, (Guid)user.AccountId, title);

            //Send a message to the All Controller
            //sectionController.server.allSection.threadController.addThread(sectionController.name, t.Id, t.title, t.comments.length, t.owner);

            MoveThreadToTop(threadId);
        }
        public void DeleteThread(ConnectedUser user, Guid sectionId, Guid threadId)
        {
            DatabaseThread.DeleteThread(DatabaseSession, (Guid) user.AccountId, sectionId, threadId);
            sectionController.MessageSender.SendRemoveThreadToAll(threadId);
            //Send a message to the All Controller
            //sectionController.server.allSection.threadController.deleteThread(sectionController.name, t.Id);
        }
        public void EditThread(ConnectedUser user, Guid sectionId, Guid threadId, String title, String description)
        {
            DatabaseThread.UpdateThreadTitle(DatabaseSession, (Guid) user.AccountId, sectionId, threadId, title);
            var (commentId, commentOwnerAccountId) = DatabaseComment.GetThreadFirstComment(DatabaseSession, threadId);
            if (commentOwnerAccountId == user.AccountId)
            {
                DatabaseComment.UpdateComment(DatabaseSession, (Guid) user.AccountId, commentId, description);
            }
            sectionController.MessageSender.SendUpdateThreadToAll(threadId, title);
            //Send a message to the All Controller
            //sectionController.server.allSection.threadController.editThread(sectionController.name, t.Id, t.title);
        }
        private void MoveThreadToTop(Guid threadId)
        {
            sectionController.MessageSender.SendMoveThreadToTopToAll(threadId);
            //Send a message to the All Controller
            //sectionController.server.allSection.threadController.moveThreadToTop(sectionController.name, t.Id);
        }
        public void AddComment(ConnectedUser user, Guid threadId, Guid sectionId, String text)
        {
            DatabaseComment.CreateComment(DatabaseSession, (Guid) user.AccountId, threadId, sectionId, text);
            // Todo: Only send comment to those viewing the thread
            //sectionController.messageSender.sendAddCommentToAll(c);

            //Send a message to the All Controller
            //sectionController.server.allSection.threadController.updateCommentCount(sectionController.name, t.Id, t.comments.length);

            MoveThreadToTop(threadId);
        }
        public void DeleteComment(ConnectedUser user, Guid commentId)
        {
            DatabaseComment.DeleteComment(DatabaseSession, (Guid) user.AccountId, commentId);
            // Todo: Only send comment delete to those viewing the thread
            //sectionController.messageSender.sendDeleteCommentToAll(c);

            //Send a message to the All Controller
            //sectionController.server.allSection.threadController.updateCommentCount(sectionController.name, t.Id, t.comments.length);
        }
        public void EditComment(ConnectedUser user, Guid commentId, string description)
        {
            DatabaseComment.UpdateComment(DatabaseSession, (Guid) user.AccountId, commentId, description);
            // Todo: Only send comment update to those viewing the thread
            //sectionController.messageSender.sendUpdateCommentToAll(c);
        }
    }
}
